datafusion/physical-expr/src/simplifier/not.rs

```diff
@@ -0,0 +1,379 @@
+
+            let new_binary = Arc::new(BinaryExpr::new(
+                left_simplified.data,
+                negated_op,
+                right_simplified.data,
+            ));
+            // We flipped the operator, so always return transformed=true
+            return Ok(Transformed::yes(new_binary));
+        }
+
+        // Handle De Morgan's laws for AND/OR
+        match binary_expr.op() {
+            Operator::And => {
+                // NOT(A AND B) -> NOT A OR NOT B
+                let not_left: Arc<dyn PhysicalExpr> =
+                    Arc::new(NotExpr::new(Arc::clone(binary_expr.left())));
+                let not_right: Arc<dyn PhysicalExpr> =
+                    Arc::new(NotExpr::new(Arc::clone(binary_expr.right())));
+
+                // Recursively simplify the NOT expressions
+                let simplified_left = simplify_not_expr_recursive(&not_left, schema)?;
+                let simplified_right = simplify_not_expr_recursive(&not_right, schema)?;
+
+                let new_binary = Arc::new(BinaryExpr::new(
+                    simplified_left.data,
+                    Operator::Or,
+                    simplified_right.data,
+                ));
+                return Ok(Transformed::yes(new_binary));
+            }
+            Operator::Or => {
+                // NOT(A OR B) -> NOT A AND NOT B
+                let not_left: Arc<dyn PhysicalExpr> =
+                    Arc::new(NotExpr::new(Arc::clone(binary_expr.left())));
+                let not_right: Arc<dyn PhysicalExpr> =
+                    Arc::new(NotExpr::new(Arc::clone(binary_expr.right())));
+
+                // Recursively simplify the NOT expressions
+                let simplified_left = simplify_not_expr_recursive(&not_left, schema)?;
+                let simplified_right = simplify_not_expr_recursive(&not_right, schema)?;
+
+                let new_binary = Arc::new(BinaryExpr::new(
+                    simplified_left.data,
+                    Operator::And,
+                    simplified_right.data,
+                ));
+                return Ok(Transformed::yes(new_binary));
+            }
+            _ => {}
+        }
+    }
+
+    // If no simplification possible, return the original expression
+    Ok(Transformed::no(expr))
+}
+
+/// Helper function that recursively simplifies expressions, including NOT expressions
+pub fn simplify_not_expr_recursive(
+    expr: &Arc<dyn PhysicalExpr>,
+    schema: &Schema,
+) -> Result<Transformed<Arc<dyn PhysicalExpr>>> {
+    // First, try to simplify any NOT expressions in this expression
+    let not_simplified = simplify_not_expr(Arc::clone(expr), schema)?;
+
+    // If the expression was transformed, we might have created new opportunities for simplification
+    if not_simplified.transformed {
+        // Recursively simplify the result
+        let further_simplified =
+            simplify_not_expr_recursive(&not_simplified.data, schema)?;
```

Comment:
I wonder whether this could be a vector of an attack.
Someone may craft a query with many `not(not(not(not(...))))` to cause a stack overflow due to the usage of recursion here.